packages/node/src/channels.ts

```diff
@@ -0,0 +1,328 @@
+  ) {}
+
+  /**
+   * Called when a new multisig is created for a set of owners.
+   * @param multisigAddress
+   * @param multisigOwners
+   * @param freeBalances
+   */
+  async createMultisig(params: Node.CreateMultisigParams): Promise<Address> {
+    const multisigAddress = Channels.getMultisigAddress(params.owners);
+    const channel: Channel = new Channel(multisigAddress, params.owners);
+    const ownersHash = Channels.canonicalizeAddresses(params.owners);
+    this.ownersToMultisigAddress[ownersHash] = multisigAddress;
+    await this.save(channel);
+    return multisigAddress;
+  }
+
+  async addMultisig(multisigAddress: Address, owners: Address[]) {
+    const channel = new Channel(multisigAddress, owners);
+    const ownersHash = Channels.canonicalizeAddresses(owners);
+    this.ownersToMultisigAddress[ownersHash] = multisigAddress;
+    await this.save(channel);
+  }
+
+  /**
+   * Returns a JSON object with the keys being the multisig addresses and the
+   * values being objects reflecting the channel schema described above.
+   */
+  async getAllChannels(): Promise<object> {
+    const channels = await this.store.get(this.multisigKeyPrefix);
+    if (!channels) {
+      console.log("No channels exist yet");
+      return {};
+    }
+    return channels;
+  }
+
+  private async getChannelFromPeerAddress(
+    peerAddress: Address
+  ): Promise<Channel> {
+    const owners = Channels.canonicalizeAddresses([
+      this.selfAddress,
+      peerAddress
+    ]);
+    const multisigAddress = this.ownersToMultisigAddress[owners];
+    const channel = await this.store.get(
+      `${this.multisigKeyPrefix}/${multisigAddress}`
+    );
+    return new Channel(
+      channel.multisigAddress,
+      channel.multisigOwners,
+      channel.appsNonce,
+      channel.freeBalances,
+      channel.appInstances,
+      channel.proposedAppInstances
+    );
+  }
+
+  public async getPeersAddressFromAppInstanceId(
+    appInstanceId: string
+  ): Promise<Address[]> {
+    const multisigAddress = this.appInstanceIdToMultisigAddress[appInstanceId];
+    const channel: Channel = await this.store.get(
+      `${this.multisigKeyPrefix}/${multisigAddress}`
+    );
+    const owners = channel.multisigOwners;
+    return owners.filter(owner => owner !== this.selfAddress);
+  }
+
+  private async getChannelFromAppInstanceId(
+    appInstanceId: string
+  ): Promise<Channel> {
+    const multisigAddress = this.appInstanceIdToMultisigAddress[appInstanceId];
+    const channel = await this.store.get(
+      `${this.multisigKeyPrefix}/${multisigAddress}`
+    );
+    return new Channel(
+      channel.multisigAddress,
+      channel.multisigOwners,
+      channel.appsNonce,
+      channel.freeBalances,
+      channel.appInstances,
+      channel.proposedAppInstances
+    );
+  }
+
+  /**
+   * Gets all appInstances across all of the channels open on this Node.
+   */
+  async getAllApps(): Promise<AppInstanceInfo[]> {
```

Comment:
@cf19drofxots We could probably have a private function that solves most of the logic and swap the key to inspect in the `channel` object, something like:

```ts
private async getAppsFrom(appsListKey: string) {
    const apps: AppInstanceInfo[] = [];
    const channels = await this.getAllChannels();
    Object.values(channels).forEach((channel: Channel) => {
      if (channel[appsListKey]) {
        apps.push(...Object.values(channel[appsListKey]));
      }
    });
    return apps;
}
```

And then the public API is reduced to:

```ts
async getAllApps(): Promise<AppInstanceInfo[]> {
  return this.getAppsFrom("appInstances");
}

async getAllPendingApps(): Promise<AppInstanceInfo[]> {
  return this.getAppsFrom("proposedAppInstances");
}
```

What do you think?

Reply:
Yeah makes sense. I hadn't gotten to retrieving the proposed app instances yet, so didn't need to generalize it just yet.